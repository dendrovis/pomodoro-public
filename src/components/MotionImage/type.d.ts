export type States = { [id: string]: { state: AnimatedType; elementContainer: HTMLDivElement } };
export type AnimatedType = 'idle' | 'run';
export type CreateContainer = (id: string) => HTMLDivElement;
export type SetAnimate = (id: string, type: AnimatedType) => void;
export interface MotionImageType {
  (id: string, elementTargetID: string): void;
  setAnimate: SetAnimate;
}
