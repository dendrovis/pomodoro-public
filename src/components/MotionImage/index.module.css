.container {
  width: 10rem;
  height: 10rem;
  filter: drop-shadow(0.3rem 0.3rem 0.2rem rgba(0 0 0 / 25%));

  & > img {
    width: clamp(50px, 100%, 500px);
  }
}
