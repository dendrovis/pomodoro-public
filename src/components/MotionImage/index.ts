import { ANIMATED_CHARACTER_IDLE, ANIMATED_CHARACTER_RUN } from '../../assets';
import { lastArrayElement } from '../../utils';
import Classes from './index.module.css';
import type { CreateContainer, MotionImageType, SetAnimate, States } from './type';

const states: States = {};

const createContainer: CreateContainer = (id) => {
  const element = document.createElement('div');
  element.id = `motion-image-${id}`;
  element.className = `${Classes.container}`;
  return element;
};

const MotionImage: MotionImageType = (id, elementTargetID) => {
  if (Object.hasOwn(states, id))
    throw new Error('existed property, make sure property name is distinct');
  const elementTarget = document.querySelector(elementTargetID);
  if (!elementTarget) throw new Error('element not available');
  const elementContainer = createContainer(id);
  states[id] = { state: 'idle', elementContainer };
  elementContainer.innerHTML = `<img src="${ANIMATED_CHARACTER_IDLE}" alt="${lastArrayElement(ANIMATED_CHARACTER_IDLE, '/')}">`;
  elementTarget.appendChild(elementContainer);
};

const setAnimate: SetAnimate = (id, type) => {
  const element = states[id].elementContainer;
  if (!element) throw new Error('element not exist');
  switch (type) {
    case 'idle':
      element.innerHTML = `<img src="${ANIMATED_CHARACTER_IDLE}" alt="${lastArrayElement(ANIMATED_CHARACTER_IDLE, '/')}">`;
      break;
    case 'run':
      element.innerHTML = `<img src="${ANIMATED_CHARACTER_RUN}" alt="${lastArrayElement(ANIMATED_CHARACTER_RUN, '/')}">`;
      break;
    default:
      throw new Error('undefined animate type');
  }
};

MotionImage.setAnimate = setAnimate;

export default MotionImage;
