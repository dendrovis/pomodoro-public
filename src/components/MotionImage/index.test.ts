import { beforeEach, beforeAll, expect, describe, it } from 'vitest';
import MotionImage from '.';

beforeEach(() => {
  document.body.innerHTML = '';
});

beforeAll(() => {
  document.body.id = 'app';
});

describe('when given id', () => {
  it('should render if distinct id', () => {
    const testID = 'test';
    MotionImage(testID, '#app');
    expect(document.body.innerHTML).toMatchSnapshot();
  });
  it('should throw error if duplicated id exist', () => {
    const testID = 'test-2';
    MotionImage(testID, '#app');
    expect(() => MotionImage(testID, '#app')).toThrowError();
  });
});

describe('when given elementTargetID', () => {
  it('should throw if element does not exist', () => {
    const testID = 'test-3';
    expect(() => MotionImage(testID, 'app')).toThrowError();
  });
  it('should render if element exist', () => {
    const testID = 'test-4';
    MotionImage(testID, '#app');
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});

describe('when setAnimate', () => {
  it('should render idle when set idle as type', () => {
    const testID = 'test-5';
    MotionImage(testID, '#app');
    MotionImage.setAnimate(testID, 'idle');
    expect(document.body.innerHTML).toMatchSnapshot();
  });
  it('should render idle when set run as type', () => {
    const testID = 'test-6';
    MotionImage(testID, '#app');
    MotionImage.setAnimate(testID, 'run');
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
