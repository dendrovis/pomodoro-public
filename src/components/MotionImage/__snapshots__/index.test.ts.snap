// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`when given elementTargetID > should render if element exist 1`] = `"<div id="motion-image-test-4" class="_container_446280"><img src="/src/assets/images/animated/idle.webp" alt="idle.webp"></div>"`;

exports[`when given id > should render if distinct id 1`] = `"<div id="motion-image-test" class="_container_446280"><img src="/src/assets/images/animated/idle.webp" alt="idle.webp"></div>"`;

exports[`when setAnimate > should render idle when set idle as type 1`] = `"<div id="motion-image-test-5" class="_container_446280"><img src="/src/assets/images/animated/idle.webp" alt="idle.webp"></div>"`;

exports[`when setAnimate > should render idle when set run as type 1`] = `"<div id="motion-image-test-6" class="_container_446280"><img src="/src/assets/images/animated/run.webp" alt="run.webp"></div>"`;
